### DISCUSSION TITLE
Using variable number of keyword arguments

### DISCUSSION THREAD

**Discussion by cve2022** [micropython/micropython#10236] [Using MicroPython] [answered]
I am running the code below from Thonny editor

```
from machine import Pin
class MyPin(Pin):
    def __init__(self, pin_nbr, *args, **kwds):
        self.pin_nbr = pin_nbr
        super().__init__(self.pin_nbr, *args, **kwds)
```

I am getting the following error (running 1.19.1 on Pico):

It works as expected...
```
>>> p4=MyPin(4, Pin.IN, Pin.PULL_DOWN)
>>> p4=MyPin(4, Pin.IN, Pin.PULL_UP)
>>> p4=MyPin(4, Pin.OUT)
>>> p4.on()
>>> p4.off()
>>> 
```
... until I add a keyword argument.

```
>>> %Run -c $EDITOR_CONTENT
>>> p4=MyPin(4, value=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 6, in __init__
TypeError: function doesn't take keyword arguments
>>> 
```
It works fine if I use the Pin class:
```
>>> p4=Pin(4, value=1)
>>> p4
Pin(4, mode=OUT)
>>> 
```
What am I missing?

**Comment by jimmo** [comment, marked as answer]
This is an interesting limitation specifically to do with how super() interacts with `__init__` of built-in types (which is called `make_new` in the C code).

Basically in the way builtin kwarg-capable functions work generally is different to `make_new` (for efficiency reasons), so there needs to be extra mapping (which isn't currently implemented).

I can't think of a good workaround, but might just be better to implement the fix.

**Reply by jimmo** [reply]
https://github.com/micropython/micropython/blob/master/py/objtype.c#L91 is the relevant code FWIW.

**Reply by cve2022** [reply]
Hello, jimmo.

Thank you for the clarification. To be honest, I just did this full inheritance experiment as a learning tool, in this particular case I really do not need the full support to Pin arguments, I only need MyPin because of the pin number attribute that I wanted to use in a MyPWM class derived from PWM as explained in the other thread. So, supporting positional arguments only is more than sufficient.

Once more, thank you very much for your help.

**Reply by ghost** [reply]
Just want to say that I ran into the same issue when implementing my own C classes. it would be nice to be able to use keyword arguments when using direct calls to `__init__`.

**Comment by peterhinch** [comment]
As a general point, subclassing of built-in types is not supported. See [this doc](https://github.com/micropython/micropython/wiki/Differences#implementation-differences). Usually you can work round this by using composition rather than inheritance.

**Reply by cve2022** [reply]
Peter, thanks for the heads up. I forgot about the implementation differences list. I always assume I am doing something wrong. In essence I am, but I assume some syntax error due to my newbie status, not that I am trying to do something that is not supported.

**Comment by stinos** [comment]
This is now implemented
